[mbstring][PHP 8.4] Add `mb_ucfirst` and `mb_lcfirst` to polyfills

Ayesh opened the pull request:
Adds polyfills for `mb_ucfirst` and `mb_lcfirst` functions based on the polyfill shown in PHP.Watch. It basically takes the first mb character, calls `mb_convert_case` with `MB_CASE_TITLE` or `MB_CASE_LOWER`, and returns with the concat of the remaining string.

The tests are taken from the php-src PR.

 - [RFC](https://wiki.php.net/rfc/mb_ucfirst)
 - [php-src PR: php-src#13161](https://github.com/php/php-src/pull/13161)
 - [PHP.Watch - PHP 8.4: New `mb_ucfirst` and `mb_lcfirst` functions](https://php.watch/versions/8.4/mb_ucfirst-mb_ucfirst)

Ayesh committed:
[mbstring][PHP 8.4] Add `mb_ucfirst` and `mb_lcfirst` to polyfills

Adds polyfills for `mb_ucfirst` and `mb_lcfirst` functions based on
the polyfill shown in PHP.Watch. It basically takes the first mb
character, calls `mb_convert_case` with `MB_CASE_TITLE` or
`MB_CASE_LOWER`, and returns with the concat of the remaining string.

The tests are taken from the php-src PR.

 - [RFC](https://wiki.php.net/rfc/mb_ucfirst)
 - [php-src PR: php-src#13161](https://github.com/php/php-src/pull/13161)
 - [PHP.Watch - PHP 8.4: New `mb_ucfirst` and `mb_lcfirst` functions](https://php.watch/versions/8.4/mb_ucfirst-mb_ucfirst)

Co-authored-by: USAMI Kenta <tadsan@zonu.me>

zonuexe commented on src/Mbstring/Mbstring.php:
```diff
@@ -871,6 +871,49 @@ public static function mb_str_pad(string $string, int $length, string $pad_strin
         }
     }
 
+    public static function mb_ucfirst(string $string, ?string $encoding = null): string {
```
It is good to add line breaks according to [Symfony Coding Standards](https://symfony.com/doc/current/contributing/code/standards.html).

```suggestion
    public static function mb_ucfirst(string $string, ?string $encoding = null): string
    {
```

zonuexe replied:
It seems good to add `mb_ucfirst` and `mb_ucfirst` to Implemented in the class header.

https://github.com/symfony/polyfill/pull/471/files#diff-3e84767d56f3132c8784f8423c914322116cef95a35b583c80070f666e9360d7R51

zonuexe commented on src/Mbstring/Mbstring.php:
```diff
@@ -871,6 +871,49 @@ public static function mb_str_pad(string $string, int $length, string $pad_strin
+    public static function mb_ucfirst(string $string, ?string $encoding = null): string {
+        if (null === $encoding) {
+            $encoding = self::mb_internal_encoding();
+        }
+
+        try {
+            $validEncoding = @self::mb_check_encoding('', $encoding);
+        } catch (\ValueError $e) {
+            throw new \ValueError(sprintf('mb_ucfirst(): Argument #2 ($encoding) must be a valid encoding, "%s" given', $encoding));
+        }
```
`self::mb_internal_encoding()` performs validation checks when setting internal values, so adding a `try-catch` to the `else` avoids a small amount of overhead. What do you think?

```suggestion
        } else {
            try {
                $validEncoding = @self::mb_check_encoding('', $encoding);
            } catch (\ValueError $e) {
                throw new \ValueError(sprintf('mb_ucfirst(): Argument #2 ($encoding) must be a valid encoding, "%s" given', $encoding));
            }
        }
```

zonuexe commented on src/Mbstring/Mbstring.php:
```diff
@@ -871,6 +871,49 @@ public static function mb_str_pad(string $string, int $length, string $pad_strin
+    public static function mb_ucfirst(string $string, ?string $encoding = null): string {
+        if (null === $encoding) {
+            $encoding = self::mb_internal_encoding();
+        }
+
+        try {
+            $validEncoding = @self::mb_check_encoding('', $encoding);
+        } catch (\ValueError $e) {
+            throw new \ValueError(sprintf('mb_ucfirst(): Argument #2 ($encoding) must be a valid encoding, "%s" given', $encoding));
+        }
+
+        // BC for PHP 7.3 and lower
+        if (!$validEncoding) {
+            throw new \ValueError(sprintf('mb_ucfirst(): Argument #2 ($encoding) must be a valid encoding, "%s" given', $encoding));
+        }
```
`self::mb_internal_encoding()` performs validation checks when setting internal values, so adding a `try-catch` to the `else` avoids a small amount of overhead. What do you think?

```suggestion
        } else {
            try {
                $validEncoding = @self::mb_check_encoding('', $encoding);
            } catch (\ValueError $e) {
                throw new \ValueError(sprintf('mb_ucfirst(): Argument #2 ($encoding) must be a valid encoding, "%s" given', $encoding));
            }
            
            // BC for PHP 7.3 and lower
            if (!$validEncoding) {
                throw new \ValueError(sprintf('mb_ucfirst(): Argument #2 ($encoding) must be a valid encoding, "%s" given', $encoding));
            }
        }
```

Ayesh replied:
You are right I understand the `else` block would avoid the non-null calls.i tried to follow the rest of the polyfill methods, so I'd prefer to keep this one like it is now.

zonuexe replied:
Perhaps this code can be condensed by combining it into a method like `self::assertEncoding()`.  Let's try another PR.

Ayesh commented:
Thanks a lot for the review @zonuexe those were great. I addressed all the points except the `else` block remark, mainly to make sure we follow the other polyfill methods, although there is indeed a merit if that changes. Perhaps it's something we can do for all methods in a separate PR?

nicolas-grekas commented:
Thank you @Ayesh.
